src/localization/features/cnn_feature.cpp

```diff
@@ -69,20 +72,15 @@ void CnnFeature::binarize() {
   }
 }
 
-double CnnFeature::computeSimilarityScore(const iFeature::ConstPtr &rhs) const {
-  const auto featurePtr = std::static_pointer_cast<const CnnFeature>(rhs);
-  if (!featurePtr) {
-    printf("[ERROR][Feature] It seems like you are trying to match features of "
-           "different type\n");
-    exit(EXIT_FAILURE);
-  }
+double CnnFeature::computeSimilarityScore(const iFeature &rhs) const {
+  const auto &feature = static_cast<const CnnFeature&>(rhs);
```

Comment:
I would say that de-facto now the code only works if you actually provide the `CnnFeature` as an `iFeature` here. If the type is different the behavior seems to be undefined. It seems like we're missing some way to work with an `iFeature` in a polymorphic manner to compare it to other features. We could do one of the following here:
- Add a method `type()` to the `iFeature` interface that would return the type of each feature object and then only perform the code in _this_ function if the types match
- Add a method `vector()` to the `iFeature` interface to ensure that we can get the values of the feature in a uniform way.

Reply:
For now, I added a type as a field in the class and the dimensions to be able to access the values from every type.